### PR TITLE
Fix/resnet take

### PR DESCRIPTION
**Pull request by robieta**
This trims the ResNet pipeline to guarantee that the number of batches during training is divisible by the number of devices. (It is not necessary during eval because eval currently only uses one device, even in MirrorStrategy) Once all_reduce is cancel-able this can be reverted.

### CODE REVIEW

File: official/resnet/resnet_run_loop.py [hunk 1]
```diff
@@ -72,6 +75,16 @@ def process_record_dataset(dataset, is_training, batch_size, shuffle_buffer,
   # dataset for the appropriate number of epochs.
   dataset = dataset.repeat(num_epochs)
 
+  if is_training and num_gpus is not None and num_gpus > 0:
```
Comment (karmel) [inline official/resnet/resnet_run_loop.py]: nit: `if is_training and num_gpus` should be sufficient.

File: official/resnet/resnet_run_loop.py [hunk 2]
```diff
@@ -72,6 +75,16 @@ def process_record_dataset(dataset, is_training, batch_size, shuffle_buffer,
...
+    total_examples = num_epochs * examples_per_epoch
+    # Force the number of batches to be divisible by the number of devices.
+    # This prevents some devices from receiving batches while others do not,
+    # which can lead to a lockup. This case will soon be handled directly by
+    # distribution strategies, at which point this .take() operation will no
+    # longer be needed.
+    total_batches = total_examples // batch_size // num_gpus * num_gpus
+    dataset.take(total_batches * batch_size)
```
Comment (karmel) [inline official/resnet/resnet_run_loop.py]: To consider: if we calculate this number in the fn that calls this one, we don't have to pass num_gpus and total_examples so far in, and can just pass an ... examples_to_use or better-named equivalent through to here.

Comment (robieta) [reply official/resnet/resnet_run_loop.py]: I considered this, but I like keeping the pipeline logic together and passing run info one layer deeper was an acceptable tradeoff.

**Comment by robieta** [comment]
I'm testing it now.

File: official/resnet/resnet_run_loop.py [hunk 3]
```diff
@@ -72,6 +75,16 @@ def process_record_dataset(dataset, is_training, batch_size, shuffle_buffer,
   # dataset for the appropriate number of epochs.
   dataset = dataset.repeat(num_epochs)
 
+  if is_training and num_gpus is not None and num_gpus > 0:
+    total_examples = num_epochs * examples_per_epoch
```
Comment (guptapriya) [inline official/resnet/resnet_run_loop.py]: do you need to also check that examples_per_epoch is not None?

Comment (robieta) [reply official/resnet/resnet_run_loop.py]: I don't believe so, because it will only be None during evaluation. But I suppose it doesn't hurt to add it.

**Pull request merged**